internal/controller/dnsrecord_healthchecks.go

```diff
@@ -2,180 +2,161 @@ package controller
 
 import (
 	"context"
-	"crypto/md5"
 	"fmt"
-	"io"
 	"reflect"
+	"strings"
 
-	"k8s.io/apimachinery/pkg/api/meta"
+	"github.com/go-logr/logr"
+
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	externaldns "sigs.k8s.io/external-dns/endpoint"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/utils/ptr"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	"github.com/kuadrant/dns-operator/api/v1alpha1"
-	"github.com/kuadrant/dns-operator/internal/provider"
+	"github.com/kuadrant/dns-operator/internal/common"
 )
 
-// healthChecksConfig represents the user configuration for the health checks
-type healthChecksConfig struct {
-	Endpoint         string
-	Port             *int64
-	FailureThreshold *int64
-	Protocol         *provider.HealthCheckProtocol
-}
-
-func (r *DNSRecordReconciler) ReconcileHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord) error {
-	var results []provider.HealthCheckResult
-	var err error
+func (r *DNSRecordReconciler) ReconcileHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord, allowInsecureCerts bool) error {
+	logger := log.FromContext(ctx).WithName("healthchecks")
+	logger.Info("Reconciling healthchecks")
 
-	dnsProvider, err := r.getDNSProvider(ctx, dnsRecord)
-	if err != nil {
-		return err
+	// Probes enabled but no health check spec yet. Nothing to do
+	if dnsRecord.Spec.HealthCheck == nil {
+		return nil
 	}
 
-	healthCheckReconciler := dnsProvider.HealthCheckReconciler()
-
-	// Get the configuration for the health checks. If no configuration is
-	// set, ensure that the health checks are deleted
-	config := getHealthChecksConfig(dnsRecord)
-
-	for _, dnsEndpoint := range dnsRecord.Spec.Endpoints {
-		addresses := provider.GetExternalAddresses(dnsEndpoint, dnsRecord)
-		for _, address := range addresses {
-			probeStatus := r.getProbeStatus(address, dnsRecord)
-
-			// no config means delete the health checks
-			if config == nil {
-				result, err := healthCheckReconciler.Delete(ctx, dnsEndpoint, probeStatus)
-				if err != nil {
-					return err
-				}
-
-				results = append(results, result)
-				continue
-			}
-
-			// creating / updating health checks
-			endpointId, err := idForEndpoint(dnsRecord, dnsEndpoint, address)
-			if err != nil {
-				return err
-			}
-
-			spec := provider.HealthCheckSpec{
-				Id:               endpointId,
-				Name:             fmt.Sprintf("%s-%s-%s", dnsRecord.Spec.RootHost, dnsEndpoint.DNSName, address),
-				Host:             &dnsRecord.Spec.RootHost,
-				Path:             config.Endpoint,
-				Port:             config.Port,
-				Protocol:         config.Protocol,
-				FailureThreshold: config.FailureThreshold,
-			}
+	desiredProbes := buildDesiredProbes(dnsRecord, common.GetLeafsTargets(common.MakeTreeFromDNSRecord(dnsRecord), ptr.To([]string{})), allowInsecureCerts)
```

Comment:
This was confusing me a little, It is fine, but I think it highlights a change needed in the tree code

```
if targets == nil{
    targets = ptr.To([]string{})
}
```

Reply:
Yes, it isn't pretty and I will revisit this part once we have it working. If this would comfort - we will never have a nil pointer with the current [implementation](https://github.com/Kuadrant/dns-operator/blob/main/internal/common/tree.go#L94)